### PR TITLE
macros: properly handles recursive macros

### LINKED ISSUE #1048: internal compiler error: in take_expression_fragment, at rust/ast/rust-ast.h:1855
```rust
macro_rules! maybe_return { ($e:expr) => ($e); }

fn frob(x: i32) -> i32{
    maybe_return! {x}
    // should return -1 
    -1
}
```

results in an ice with the version of gccrs on compiler explorer (1bb9a29688ab4ddfec7f8d36ca2cee63c5f258d2).

```
rust1: internal compiler error: in take_expression_fragment, at rust/ast/rust-ast.h:1855
0x2042789 internal_error(char const*, ...)
	???:0
0x80023f fancy_abort(char const*, int, char const*)
	???:0
0x92a158 Rust::AST::ASTFragment::take_expression_fragment()
	???:0
0x9271f4 Rust::AttrVisitor::visit(Rust::AST::ArithmeticOrLogicalExpr&)
	???:0
0x9283a5 Rust::AttrVisitor::visit(Rust::AST::BlockExpr&)
	???:0
0x9268be Rust::AttrVisitor::visit(Rust::AST::Function&)
	???:0
0x8d27fe Rust::MacroExpander::expand_crate()
	???:0
0x8bec2e Rust::Session::expansion(Rust::AST::Crate&)
	???:0
0x8c1de4 Rust::Session::parse_file(char const*)
	???:0
0x8c29c0 Rust::Session::parse_files(int, char const**)
	???:0
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
Compiler returned: 1
```

_Originally posted by @bjorn3 in https://github.com/Rust-GCC/gccrs/issues/22#issuecomment-1074171728_

### LINKED ISSUE #1084: macros: Any macro in the return position will result in Rust GCC stuck
I tried this code:

```rust
macro_rules! add {
    ($a:expr) => { $a };
    ($a:expr, $($b:expr),+) => { $a + add!($($b),*) }
}

fn test() -> i32 {
    add!(add!(1, 2), add!(1, 2))
}

fn main() {
    let _ = test();
}
```

A reduced case for `gccrs`:

```rust
macro_rules! add {
    ($a:expr) => { $a };
    ($a:expr, $($b:expr),+) => { $a + add!($($b),*) }
}

fn main() -> i32 {
    add!(add!(1, 2))
}
```

I expected to see this happen: `gccrs` should compile the code, and the generated program should run and exit with `6` (the reduced case should exit with `3`).

Instead, this happened: `gccrs` stuck in an infinite loop in the macro expansion pass.

### Stack Trace

<details>
<summary>Click to show detailed stack trace</summary>
<p>

```
#0  0x00007ffff79cf4c4 in __memmove_avx_unaligned_erms () from /usr/lib/libc.so.6
#1  0x00000000020f5731 in std::basic_streambuf<char, std::char_traits<char> >::xsputn(char const*, long) ()
#2  0x00000000020e5f54 in std::basic_ostream<char, std::char_traits<char> >& std::__ostream_insert<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*, long) ()
#3  0x0000000000934a03 in std::operator<< <std::char_traits<char> > (__c=<optimized out>, __out=...) at /usr/include/c++/11.2.0/ostream:518
#4  expand_format (fmt=<optimized out>) at ../../gcc/rust/rust-diagnostics.cc:42
#5  expand_message(const char *, typedef __va_list_tag __va_list_tag *) (fmt=<optimized out>, ap=0x7fffffffc788) at ../../gcc/rust/rust-diagnostics.cc:114
#6  0x00000000009353b2 in Rust::Error::Error (this=this@entry=0x7fffffffc960, location=..., fmt=fmt@entry=0x2139080 "unrecognised token %qs for start of %s")
    at ../../gcc/rust/rust-diagnostics.cc:239
#7  0x0000000000a50d65 in Rust::Parser<Rust::MacroInvocLexer>::parse_item (this=0x7fffffffcb80, called_from_statement=called_from_statement@entry=true)
    at ../../gcc/rust/parse/rust-parse-impl.h:1095
#8  0x0000000000a059b9 in operator() (__closure=<optimized out>, __closure=<optimized out>) at ../../gcc/rust/expand/rust-macro-expand.cc:811
#9  std::__invoke_impl<Rust::AST::SingleASTNode, Rust::transcribe_many_items(Rust::Parser<Rust::MacroInvocLexer>&, Rust::TokenId&)::<lambda()>&> (__f=...)
    at /usr/include/c++/11.2.0/bits/invoke.h:61
#10 std::__invoke_r<Rust::AST::SingleASTNode, Rust::transcribe_many_items(Rust::Parser<Rust::MacroInvocLexer>&, Rust::TokenId&)::<lambda()>&> (__fn=...)
    at /usr/include/c++/11.2.0/bits/invoke.h:143
#11 std::_Function_handler<Rust::AST::SingleASTNode(), Rust::transcribe_many_items(Rust::Parser<Rust::MacroInvocLexer>&, Rust::TokenId&)::<lambda()> >::_M_invoke(const std::_Any_data &) (
    __functor=...) at /usr/include/c++/11.2.0/bits/std_function.h:292
#12 0x00000000009fbed4 in std::function<Rust::AST::SingleASTNode ()>::operator()() const (this=0x7fffffffcc10) at /usr/include/c++/11.2.0/bits/std_function.h:560
#13 Rust::parse_many(Rust::Parser<Rust::MacroInvocLexer> &, Rust::TokenId &, std::function<Rust::AST::SingleASTNode()>) (parser=..., delimiter=@0x7fffffffcab0: Rust::RIGHT_CURLY, 
    parse_fn=...) at ../../gcc/rust/expand/rust-macro-expand.cc:794
#14 0x0000000000a041f3 in Rust::transcribe_many_stmts (delimiter=<optimized out>, parser=...) at ../../gcc/rust/expand/rust-macro-expand.cc:896
#15 Rust::transcribe_on_delimiter (last_token_id=<optimized out>, delimiter=<optimized out>, semicolon=<optimized out>, parser=...) at ../../gcc/rust/expand/rust-macro-expand.cc:930
#16 Rust::transcribe_context (last_token_id=<optimized out>, delimiter=<optimized out>, semicolon=false, parser=..., ctx=Rust::MacroExpander::ITEM)
    at ../../gcc/rust/expand/rust-macro-expand.cc:978
#17 Rust::MacroExpander::transcribe_rule (this=<optimized out>, match_rule=..., invoc_token_tree=..., matched_fragments=..., semicolon=<optimized out>, ctx=Rust::MacroExpander::ITEM)
    at ../../gcc/rust/expand/rust-macro-expand.cc:1054
#18 0x0000000000a07573 in Rust::MacroExpander::expand_decl_macro (this=0x7fffffffd100, invoc_locus=..., invoc=..., rules_def=..., semicolon=<optimized out>)
    at ../../gcc/rust/expand/rust-macro-expand.cc:149
#19 0x0000000000a078db in Rust::MacroExpander::expand_invoc (this=0x7fffffffd100, invoc=...) at ../../gcc/rust/expand/rust-macro-expand.cc:205
#20 0x0000000000a56b75 in Rust::AttrVisitor::visit (this=0x7fffffffcfe0, expr=...) at ../../gcc/rust/expand/rust-attribute-visitor.cc:622
#21 0x00000000009fc43f in Rust::MacroExpander::take_expanded_fragment (vis=..., this=0x7fffffffd100) at ../../gcc/rust/expand/rust-macro-expand.h:303
#22 Rust::MacroExpander::expand_crate (this=this@entry=0x7fffffffd100) at ../../gcc/rust/expand/rust-macro-expand.cc:382
#23 0x00000000009dc376 in Rust::Session::expansion (this=this@entry=0x2bf56c0 <Rust::Session::get_instance()::instance>, crate=...) at ../../gcc/rust/rust-session-manager.cc:831
#24 0x00000000009e888a in Rust::Session::parse_file (this=this@entry=0x2bf56c0 <Rust::Session::get_instance()::instance>, filename=<optimized out>)
    at ../../gcc/rust/rust-session-manager.cc:568
#25 0x00000000009e8e31 in Rust::Session::parse_files (this=0x2bf56c0 <Rust::Session::get_instance()::instance>, num_files=1, files=<optimized out>)
    at ../../gcc/rust/rust-session-manager.cc:484
#26 0x00000000010a0dc1 in compile_file () at ../../gcc/toplev.cc:452
#27 0x000000000092c96e in do_compile (no_backend=false) at ../../gcc/toplev.cc:2158
#28 toplev::main (this=this@entry=0x7fffffffd876, argc=<optimized out>, argc@entry=6, argv=<optimized out>, argv@entry=0x7fffffffd9a8) at ../../gcc/toplev.cc:2310
#29 0x000000000092e8ec in main (argc=6, argv=0x7fffffffd9a8) at ../../gcc/main.cc:39
```
</p>
</details>

### Meta

- Rust GCC version: 9011184f38a04f81ba3194b826bec3f30a11c07b

### PR DESCRIPTION
**Pull request by liushuyu**
- macros: properly handles recursive macros

Fixes #1084

### CODE REVIEW

File: gcc/rust/expand/rust-attribute-visitor.cc [hunk 1]
```diff
@@ -402,10 +402,7 @@ AttrVisitor::visit (AST::MacroInvocation &macro_invoc)
   // I don't think any macro token trees can be stripped in any way
 
   // TODO: maybe have cfg! macro stripping behaviour here?
-  if (macro_invoc.has_semicolon ())
-    expander.expand_invoc_semi (macro_invoc);
-  else
-    expander.expand_invoc (macro_invoc);
+  expander.expand_invoc (macro_invoc, macro_invoc.has_semicolon ());
```
Comment (CohenArthur) [inline gcc/rust/expand/rust-attribute-visitor.cc:405]: That's much better than what was here, thank you. I think there's a FIXME somewhere dedicated to this but it's not important.

File: gcc/rust/expand/rust-attribute-visitor.cc [hunk 2]
```diff
@@ -3434,8 +3431,17 @@ void
...
+  while (fragment.should_expand ())
+    {
+      expr = fragment.take_expression_fragment ();
+      expander.expansion_depth++;
+      auto new_fragment = expander.take_expanded_fragment (*this);
+      if (new_fragment.is_error ())
+	break;
+      fragment = std::move (new_fragment);
+    }
+  expander.expansion_depth = original_depth;
```
Comment (CohenArthur) [inline gcc/rust/expand/rust-attribute-visitor.cc:3444]: That's really interesting, I didn't think we needed that. How come this is only necessary for expression fragments?

**Review by CohenArthur — APPROVED** [review]
I think this is great work. Thank you for working on this. I hope the code wasn't too messy to understand! Sorry about that!

File: gcc/rust/expand/rust-macro-expand.cc [hunk 3]
```diff
@@ -136,54 +136,19 @@ MacroExpander::expand_invoc (AST::MacroInvocation &invoc)
...
+  if (has_semicolon)
+    source_node = invoc.get_macro_node_id ();
   else
-    fragment
-      = expand_decl_macro (invoc.get_locus (), invoc_data, *rules_def, false);
-
-  set_expanded_fragment (std::move (fragment));
-}
-
-// FIXME: Arthur: Refactor these two functions, they're really similar
```
Comment (liushuyu) [inline gcc/rust/expand/rust-macro-expand.cc:166]: Are you looking for this `FIXME` comment? 😄

File: gcc/rust/expand/rust-attribute-visitor.cc (hunk 2 above)
Comment (liushuyu) [reply gcc/rust/expand/rust-attribute-visitor.cc:3444]: > That's really interesting, I didn't think we needed that. How come this is only necessary for expression fragments?

I am not very sure either. But when I was debugging the issue, I noticed the fragment was not expanded completely in this step and caused the parser to stuck.

File: gcc/rust/expand/rust-macro-expand.cc (hunk 3 above)
Comment (CohenArthur) [reply gcc/rust/expand/rust-macro-expand.cc:166]: Perfect hahaha, thank you :)

**Review by philberty — APPROVED** [review]
LGTM @CohenArthur is best placed to merge this one

**Comment by CohenArthur** [comment]
Oh right, sorry!


bors r+

**Pull request merged**